### PR TITLE
ISPN-15172 SIFS throws exception if told to compact a partially writt…

### PR DESCRIPTION
**Pull request by wburns**
…en entry

https://issues.redhat.com/browse/ISPN-15172

### CODE REVIEW

**Comment by wburns** [comment]
Now the compactor will check the header to ensure there is enough space left to read the amount of bytes in the file. If it does not it will just ignore that entry and log a WARN message. If the file was not written to the log appender then it cannot be in the index, so it is effectively a lost entry (the user may or may not have gotten an error back from the original request).

Also after the file size has been guaranteed I changed the error messages if for some reason the data file is concurrently changed to no longer have enough bytes that the header has and will throw an exception for this as it is a bug of some type.

File: core/src/main/java/org/infinispan/persistence/sifs/Compactor.java [hunk 1]
```diff
@@ -466,18 +470,34 @@ private void compactSingleFile(int scheduledFile, boolean isLogFile, CompactionE
...
             if (header == null) {
                break;
             }
+            long remainingBytes = fileSize - scheduledOffset;
+            if (header.totalLength() > remainingBytes) {
+               // Attempt to read the key to give a better warning
+               if (header.keyLength() > remainingBytes) {
+                  byte[] serializedKey = EntryRecord.readKey(handle, header, scheduledOffset);
+                  if (serializedKey != null) {
+                     log.warnf("Compactor encountered a corrupted entry for key %s at %s:%s that does not contain enough bytes, dropping entry! This was probably caused by unexpected process stoppage prior",
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/sifs/Compactor.java:487]: Oops forgot to add these with i18n.

File: core/src/main/java/org/infinispan/persistence/sifs/Compactor.java [hunk 2]
```diff
@@ -466,18 +470,30 @@ private void compactSingleFile(int scheduledFile, boolean isLogFile, CompactionE
...
          while (!clearSignal.get() && !terminateSignal) {
             EntryHeader header = EntryRecord.readEntryHeader(handle, scheduledOffset);
             if (header == null) {
                break;
             }
+            long remainingBytes = fileSize - scheduledOffset;
+            if (header.totalLength() > remainingBytes) {
+               byte[] serializedKey = null;
+               // Attempt to read the key to give a better warning
+               if (header.keyLength() > remainingBytes) {
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/sifs/Compactor.java:485]: Ah, yes. Good catch, it wasn't printing the key then :)

File: core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreCorruptRestartTest.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+      // Force compaction for the previous file
+      CompletionStages.join(compactor.forceCompactionForAllNonLogFiles());
+
+      assertEquals(1, cache(0, cacheName).size());
+   }
+
+   MarshallableEntry objectToMarshalledBuffer(Object obj) {
+      DataConversion dc = cache(0, cacheName).getAdvancedCache().getKeyDataConversion();
+      Object storedKey = dc.toStorage("key-1");
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreCorruptRestartTest.java:103]: Yeah, that was my half attempt to refactor and one point and forgot to update :)

**Comment by wburns** [comment]
Updated to address comments.

**Review by jabolina — APPROVED** [review]

**Comment by jabolina** [comment]
Thanks, @wburns!

**Pull request merged**